### ISSUE TITLE
Programmatically validate object against collection config

### ISSUE THREAD

**Issue by mmkal** [netlify/netlify-cms#5955]
**Is your feature request related to a problem? Please describe.**

In certain situations, it's useful to be able to validate an object against a collection config. For example, when the config is changed (e.g. a required field added), a CI job could use a validation API to detect content that is no longer in the right format.

**Describe the solution you'd like**

Something like this:

```js
const json = fs.readFileSync('path/to/content.json').toString()
const obj = JSON.parse(json)
const collection = config.collections.find(c => c.name = 'foo')
CMS.validate(collection, obj) // throws if obj doesn't conform to expected shape
```

**Possible implementation**

It may be achievable via a static function:

```js
const fieldToJsonSchema = field => {
  if (field.widget === 'string') {
    return { type: 'string' }
  }
  ...

  if (field.widget === 'object') {
    const schema = {
      properties: Object.fromEntries(
        field.fields.map(f => [f.name, fieldToJsonSchema(f)]),
      ),
      required: field.fields.map(f => f.name).filter(f => f.required !== true),
  }

  if (field.widget === 'list') {
    // ...
  }
}

CMS.validate = (collection, obj) => {
  return someJsonSchemaValidatorFromNpm.validate(fieldToJsonSchema(collection), obj)
}
```

There are a fair few edge-cases though.

**Describe alternatives you've considered**

N/A.

**Additional context**

Originally asked in Slack: https://netlifycms.slack.com/archives/CPRR0PQ9E/p1636036471070400

___

Note: I'd be willing to work on this if it would be accepted.

**Comment by erezrokah** [comment]
@mmkal thanks for opening this issue and suggesting an approach.

Users can register their own custom widgets, so the CMS core would not know how to handle those.

As a general approach the CMS delegates handling data to the widgets, as some widgets has logic on how to handle the data based on the widget configuration. For example a datetime widget can have multiple formats, and a list widget can have variable types. The relation widget can be multiple selection, or single selection and so on.

**Comment by mmkal** [comment]
Absolutely, the devil would be in the details for `fieldToJsonSchema`. Complicated widgets like datetime widgets would have correspondingly complicated implementations, but still doable I think. Potentially problematic to need to mirror all widget behavior separately though. Maybe a solution that would work for custom widgets 
too would be to introduce an (optional) `toJsonSchema` function on every widget? If not supplied, it could default to `{}` (json-schema's equivalent of `any`).
